### PR TITLE
[Clock] Documentation for new Clock component

### LINKED ISSUE #17074: [Clock] A new component to decouple applications from the system clock
| Q            | A
| ------------ | ---
| Feature PR   | symfony/symfony#46715
| PR author(s) | @nicolas-grekas
| Merged in    | 6.2

### PR DESCRIPTION
**Pull request by dbrumann**
Closes #17074

### CODE REVIEW

File: components/clock.rst [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+.. _`clock-component`:
+
+The Clock Component
+===================
+
+The Clock component decouples applications from the system clock. This improves
+testability of time-sensitive logic.
+
+The component provides a ``ClockInterface`` with 3 different clock
```
Comment (OskarStark) [inline components/clock.rst]:
```suggestion
The component provides a ``ClockInterface`` with the following clock
```

File: components/clock.rst [hunk 2]
```diff
@@ -0,0 +1,88 @@
+.. index::
+   single: Clock
+   single: Components; Clock
+
+.. _`clock-component`:
+
+The Clock Component
+===================
+
```
Comment (OskarStark) [inline components/clock.rst:7]: `versionadded` is missing

File: components/clock.rst [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    $ composer require symfony/clock
+
+.. include:: /components/require_autoload.rst.inc
+
+Usage
+-----
+
+A clock service replaces creating a new ``DateTime`` or
+``DateTimeImmutable``-object for the current time. Instead, you inject the
```
Comment (OskarStark) [inline components/clock.rst]:
```suggestion
``DateTimeImmutable`` object for the current time. Instead, you inject the
```

File: components/clock.rst [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+``ClockInterface`` and call ``now()``. By default, your application will likely
+use a ``NativeClock``, which always returns the current system time. In tests it is replaced by a ``MockClock``. The
+``MockClock`` is instantiated with a time and does not move forward on its own, meaning when you create an
+instance with a time, it is always kept, unless you change it using
+either ``sleep()`` or ``modify()``. This gives you full control over what your
+code assumes is the current time.
+
+The following example introduces a service utilizing the clock component to
+determine the current time.
```
Comment (OskarStark) [inline components/clock.rst]:
```suggestion
determine the current time::
```

File: components/clock.rst [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        ) {}
+
+        public function isExpired(DateTimeInterface $time): bool
+        {
+            return $this->clock->now() > $time;
+        }
+    }
+
+When writing a test for this service, you can check both cases where something
+is expired or not, by modifying the clock's time.
```
Comment (OskarStark) [inline components/clock.rst]:
```suggestion
is expired or not, by modifying the clock's time::
```

File: components/clock.rst [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    use Symfony\Component\Clock\MockClock;
```
Comment (OskarStark) [inline components/clock.rst]: Lets add a `use` for `TestCase`

File: components/clock.rst [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        public function testIsExpired(): void
+        {
+            $clock = new MockClock('2022-11-16 15:20:00');
+            $expirationChecker = new ExpirationChecker($clock);
+            $time = new DateTimeImmutable('2022-11-16 15:25:00');
+
+            // $time is in the future, so it is not expired
+            static::assertFalse($expirationChecker->isExpired($time));
+
+            // Clock sleeps for 10min, so now is '2022-11-16 15:30:00'
```
Comment (OskarStark) [inline components/clock.rst]:
```suggestion
            // Clock sleeps for 10 minutes, so now is '2022-11-16 15:30:00'
```

File: components/clock.rst [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            $clock->sleep(600); // Instantly changes time as if we waited for 10min (600secs)
```
Comment (OskarStark) [inline components/clock.rst]:
```suggestion
            $clock->sleep(600); // Instantly changes time as if we waited for 10 minutes (600secs)
```

File: components/clock.rst [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        public function __construct(
+            private readonly ClockInterface $clock
+        ) {}
+
+        public function isExpired(DateTimeInterface $time): bool
+        {
+            return $this->clock->now() > $time;
+        }
+    }
+
```
Comment (OskarStark) [inline components/clock.rst:61]: Could have its own headline IMO

**Review by OskarStark — COMMENTED** [review]
👍

File: components/clock.rst (hunk 9 above)
Comment (dbrumann) [reply components/clock.rst:61]: I'm sorry, I don't understand? What headline do you have in mind?

Comment (OskarStark) [reply components/clock.rst:61]:
```suggestion
Testing
-------
```
👍

**Comment by OskarStark** [comment]
The Code-Blocks stuff can be ignored

File: components/clock.rst [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+.. versionadded:: 6.2
+
+    The Clock component was introduced in Symfony 6.2
+
+The Clock component decouples applications from the system clock. This improves
+testability of time-sensitive logic.
+
+The component provides a ``ClockInterface`` with the following implementations
+for different use cases.
```
Comment (xabbuh) [inline components/clock.rst]:
```suggestion
for different use cases:
```

File: components/clock.rst [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+* ``NativeClock`` provides a way to interact with the system clock.
```
Comment (xabbuh) [inline components/clock.rst]:
```suggestion
* ``NativeClock`` provides a way to interact with the system clock;
```

File: components/clock.rst [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+* ``MockClock`` is commonly used in tests as a replacement for the
+  ``NativeClock`` to be able to freeze and change the current time using either
+  ``sleep()`` or ``modify()``.
```
Comment (xabbuh) [inline components/clock.rst]:
```suggestion
  ``sleep()`` or ``modify()``;
```

File: components/clock.rst [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+.. include:: /components/require_autoload.rst.inc
+
+Usage
+-----
+
+A clock service replaces creating a new ``DateTime`` or
+``DateTimeImmutable`` object for the current time. Instead, you inject the
+``ClockInterface`` and call ``now()``. By default, your application will likely
+use a ``NativeClock``, which always returns the current system time. In tests it is replaced by a ``MockClock``. The
```
Comment (xabbuh) [inline components/clock.rst]: replaced *with*?

File: components/clock.rst [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+``MockClock`` is instantiated with a time and does not move forward on its own, meaning when you create an
+instance with a time, it is always kept, unless you change it using
+either ``sleep()`` or ``modify()``. This gives you full control over what your
```
Comment (xabbuh) [inline components/clock.rst]: The `MockClock` is instantiated with a time and does not move forward on its own. The time is fixed until `sleep()` or `modify()` are called.

File: components/clock.rst [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+code assumes is the current time.
+
+The following example introduces a service utilizing the clock component to
```
Comment (xabbuh) [inline components/clock.rst]:
```suggestion
The following example introduces a service utilizing the Clock component to
```

File: components/clock.rst [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+    use Symfony\Component\Clock\ClockInterface;
+
+    class ExpirationChecker
+    {
+        public function __construct(
+            private readonly ClockInterface $clock
+        ) {}
+
+        public function isExpired(DateTimeInterface $time): bool
```
Comment (xabbuh) [inline components/clock.rst]: Maybe rename `$time` to `$validUntil`?

File: components/clock.rst [hunk 17]
```diff
@@ -0,0 +1,96 @@
+.. index::
+   single: Clock
+   single: Components; Clock
+
+.. _`clock-component`:
```
Comment (wouterj) [inline components/clock.rst]:
```suggestion
```

(this system is a custom anchor - but we don't use it, so let's remove it)

**Review by javiereguiluz — APPROVED** [review]
@dbrumann thanks a lot for contributing these docs!

I like your proposed contents a lot and I don't have any comment/tweak to do about them 🙂 My only comment is about `MonotonicClock`, which is not explained in these docs and I don't fully understand it.

But maybe we can leave it for a future PR. Thanks.

File: components/clock.rst [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+
+The Clock Component
+===================
+
+.. versionadded:: 6.2
+
+    The Clock component was introduced in Symfony 6.2
+
+The Clock component decouples applications from the system clock. This improves
+testability of time-sensitive logic.
```
Comment (wouterj) [inline components/clock.rst]:
```suggestion
The Clock component decouples applications from the system clock. This allows
you to fix time to improve testability of time-sensitive logic.
```

just a suggestion, but I think we should clarify a bit on the *why* about improving testability

File: components/clock.rst [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+The component provides a ``ClockInterface`` with the following implementations
+for different use cases:
+
+* ``NativeClock`` provides a way to interact with the system clock;
+* ``MockClock`` is commonly used in tests as a replacement for the
+  ``NativeClock`` to be able to freeze and change the current time using either
+  ``sleep()`` or ``modify()``;
+* ``MonotonicClock`` relies on ``hrtime()`` and provides a high resolution,
+  monotonic clock, when you need a precise stopwatch.
```
Comment (wouterj) [inline components/clock.rst]:
```suggestion
:class:`Symfony\\Component\\Clock\\NativeClock`
    Provides a way to interact with the system clock, this is the same as doing
    ``new \DateTimeImmutable()``.
:class:`Symfony\\Component\\Clock\\MockClock`
    Commonly used in tests as a replacement for the ``NativeClock`` to be able
    to freeze and change the current time using either ``sleep()`` or ``modify()``.
:class:`Symfony\\Component\\Clock\\MonotonicClock``
    Relies on ``hrtime()`` and provides a high resolution, monotonic clock,
    when you need a precise stopwatch.
```

definition lists for the win :)

File: components/clock.rst [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+            // $validUntil is in the future, so it is not expired
+            static::assertFalse($expirationChecker->isExpired($validUntil));
+
+            // Clock sleeps for 10 minutes, so now is '2022-11-16 15:30:00'
+            $clock->sleep(600); // Instantly changes time as if we waited for 10 minutes (600secs)
+
+            // $validUntil is in the past after advancing the clock, so it is expired
+            static::assertTrue($expirationChecker->isExpired($validUntil));
+
+            $clock->modify('2022-11-16 15:00:00');
```
Comment (wouterj) [inline components/clock.rst:92]:
```suggestion
            // modify the clock, accepts all formats supported by DateTimeImmutable::modify()
            $clock->modify('2022-11-16 15:00:00');
```

File: components/clock.rst [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+Usage
+-----
+
+A clock service replaces creating a new ``DateTime`` or
+``DateTimeImmutable`` object for the current time. Instead, you inject the
+``ClockInterface`` and call ``now()``. By default, your application will likely
+use a ``NativeClock``, which always returns the current system time. In tests it is replaced with a ``MockClock``. The
+The ``MockClock`` is instantiated with a time and does not move forward on its own. The time is
+fixed until ``sleep()`` or ``modify()`` are called. This gives you full control over what your code
+assumes is the current time.
```
Comment (wouterj) [inline components/clock.rst]: I think we should move the mock clock information the testing section below (and rename it to "Testing using MockClock"?). Then we can also explain that sleep doesn't actually sleep using the mock clock.

**Review by wouterj — COMMENTED** [review]
Thank you Denis!

Sorry to be the third doc member adding lots of annotations. I'm also fine with doing the changes while merging :)

**Comment by dbrumann** [comment]
No problem. I will work on the changes tomorrow during my train ride. Keep them coming.

**Comment by dbrumann** [comment]
@javiereguiluz I added a new section describing the purpose of the monotonic clock. Unfortunately, I couldn't come up with a relevant code snippet but maybe we don't need one or someone else has a good idea?

**Comment by dbrumann** [comment]
@javiereguiluz I added a new section describing the purpose of the monotonic clock. Unfortunately, I couldn't come up with a relevant code snippet but maybe we don't need one or someone else has a good idea?

File: components/clock.rst [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+        }
+    }
+
+Monotonic Clock
+---------------
+
+The ``MonotonicClock`` allows you to implement a precise stopwatch, depending on the system up to
+nanosecond precision. It can be used to measure the elapsed time between 2 calls without being
+affected by inconsistencies sometimes introduced by the system clock, e.g. by updating it. Instead,
+it consistently increases time, making it especially useful for measuring performance consistently.
```
Comment (OskarStark) [inline components/clock.rst]: consistently ist used twice here

**Comment by javiereguiluz** [comment]
This is now merged 🥳 

Denis, thanks a lot for this nice contribution! And thanks to our great reviewers too.

**Pull request merged**